src/compiler/transformers/generators.ts

```diff
@@ -772,6 +774,65 @@ namespace ts {
             return visitEachChild(node, visitor, context);
         }
 
+        /**
+         * Visits a comma expression containing `yield`.
+         *
+         * @param node The node to visit.
+         */
+        function visitCommaExpression(node: BinaryExpression) {
+            // [source]
+            //      x = a(), yield, b();
+            //
+            // [intermediate]
+            //      a();
+            //  .yield resumeLabel
+            //  .mark resumeLabel
+            //      x = %sent%, b();
+
+            let pendingExpressions: Expression[] = [];
+            visit(node.left);
+            visit(node.right);
+            return factory.inlineExpressions(pendingExpressions);
+
+            function visit(node: Expression) {
+                if (isBinaryExpression(node) && node.operatorToken.kind === SyntaxKind.CommaToken) {
+                    visit(node.left);
+                    visit(node.right);
+                }
+                else {
+                    if (containsYield(node) && pendingExpressions.length > 0) {
+                        emitWorker(OpCode.Statement, [factory.createExpressionStatement(factory.inlineExpressions(pendingExpressions))]);
+                        pendingExpressions = [];
+                    }
+
+                    pendingExpressions.push(visitNode(node, visitor, isExpression));
+                }
+            }
+        }
+
+        /**
+         * Visits a comma-list expression.
+         *
+         * @param node The node to visit.
+         */
+        function visitCommaListExpression(node: CommaListExpression) {
+            // flattened version of `visitCommaExpression`
+            let pendingExpressions: Expression[] = [];
+            for (const elem of node.elements) {
+                if (isBinaryExpression(elem) && elem.operatorToken.kind === SyntaxKind.CommaToken) {

```

Comment:
Because of the force-push I couldn't tell whether my previous suggestion changed anything in the tests (to be fair, it might be difficult to get into that state since the only place that creates `CommaListExpression` flattens existing commas). The only way I could see that happening is with a custom "before" transform", so its good to keep this in even if we may never actually hit this case.